### PR TITLE
#20 - Add GET /department by businessId endpoint

### PR DESCRIPTION
**Pull request by harpestes**

### CODE REVIEW

File: src/main/java/com/gachigang/ontherun/service/DepartmentService.java [hunk 1]
```diff
@@ -2,24 +2,33 @@
...
     private final DepartmentRepository departmentRepository;
     private final UserRepository userRepository;
+    private final BusinessRepository businessRepository;
 
     @Transactional(readOnly = true)
     public Department getDepartmentByUserId(@NonNull final Long id) {
         return departmentRepository.findDepartmentByUsersContains(userRepository.findById(id)
                 .orElseThrow(NotFoundException::new));
     }
+
```
Comment (Konoplastiy) [inline src/main/java/com/gachigang/ontherun/service/DepartmentService.java:29]: Add @transactional(readonly = true)

File: src/main/java/com/gachigang/ontherun/service/DepartmentService.java [hunk 2]
```diff
@@ -2,24 +2,33 @@
...
+    public Set<Department> getDepartmentsByBusinessId(@NonNull Long businessId) {
+        return departmentRepository.findDepartmentsByBusiness(businessRepository.findById(businessId)
+                .orElseThrow(() -> new IllegalArgumentException("Business with id " + businessId + " not found")));
```
Comment (Konoplastiy) [inline src/main/java/com/gachigang/ontherun/service/DepartmentService.java:32]: I guess, it should throw NotFoundException instead of illegalargumentexception
@AlexandrVikhristyk What do u think about it?

File: src/test/java/com/gachigang/ontherun/controller/DepartmentControllerTest.java [hunk 3]
```diff
@@ -20,12 +21,21 @@ public class DepartmentControllerTest {
...
-    void getDepartmentByUserId() {
+    void testGetDepartmentByUserId() {
         User user = User.builder()
                         .id(1L)
                         .lastname("test")
                         .build();
         departmentController.getDepartmentByUserId(user);
         verify(departmentService, times(1)).getDepartmentByUserId(user.getId());
     }
+
```
Comment (Konoplastiy) [inline src/test/java/com/gachigang/ontherun/controller/DepartmentControllerTest.java:32]: The test not bad but we need to test all variants like positive and negative

File: src/main/java/com/gachigang/ontherun/service/DepartmentService.java (hunk 2 above)
Comment (AlexandrVikhristyk) [reply src/main/java/com/gachigang/ontherun/service/DepartmentService.java:32]: Yep, I also think it whould be better to change to NotFoundException

File: src/test/java/com/gachigang/ontherun/controller/DepartmentControllerTest.java (hunk 3 above)
Comment (harpestes) [reply src/test/java/com/gachigang/ontherun/controller/DepartmentControllerTest.java:34]: Hope, i've done it

File: src/main/java/com/gachigang/ontherun/controller/DepartmentController.java [hunk 4]
```diff
@@ -28,4 +30,16 @@ public class DepartmentController {
...
+
+    /**
+     * Endpoint for getting the departments of the specific business via id
+     *
+     * @return A list of {@link Department} objects.
+     * @throws {@link NotFoundException} if business with id was not found,
+     * or {@link RuntimeException} if there is no departments for this business
+     */
+    @GetMapping("/{businessId}")
+    public Set<Department> getDepartmentsByBusinessId(@PathVariable Long businessId) {
```
Comment (AlexandrVikhristyk) [inline src/main/java/com/gachigang/ontherun/controller/DepartmentController.java:42]: Hmmm, we have to return DTO from controller layer, even we must -- because otherway we will get recurcive JSON in responce and a lot of queries in DB

File: src/main/java/com/gachigang/ontherun/service/DepartmentService.java [hunk 5]
```diff
@@ -2,24 +2,33 @@
...
     @Transactional(readOnly = true)
     public Department getDepartmentByUserId(@NonNull final Long id) {
         return departmentRepository.findDepartmentByUsersContains(userRepository.findById(id)
                 .orElseThrow(NotFoundException::new));
     }
+
+    @Transactional(readOnly = true)
+    public Set<Department> getDepartmentsByBusinessId(@NonNull Long businessId) {
+        return departmentRepository.findDepartmentsByBusiness(businessRepository.findById(businessId)
+                .orElseThrow(NotFoundException::new));
```
Comment (AlexandrVikhristyk) [inline src/main/java/com/gachigang/ontherun/service/DepartmentService.java:32]: I like this approach, lets try to follow same way throught whole project

File: src/test/java/com/gachigang/ontherun/controller/DepartmentControllerTest.java [hunk 6]
```diff
@@ -20,12 +23,46 @@ public class DepartmentControllerTest {
...
+    @Test
+    void testGetDepartmentsByBusinessIdValidId() {
```
Comment (AlexandrVikhristyk) [inline src/test/java/com/gachigang/ontherun/controller/DepartmentControllerTest.java:36]: Honestly it depends on convenience, but foe mt it is hard to read huge camel case text, there are up to 10 naming convections, I suggest simply split main text by underscore -- for instance  testGetDepartmentsByBusinessId_ValidId -- so the first part tells us the name of testing method and the second its condition

Comment (AlexandrVikhristyk) [inline src/test/java/com/gachigang/ontherun/controller/DepartmentControllerTest.java:36]: Oh, sorry I guess @Konoplastiy and I have missed it :( In controller we have only one method call and 0 bussines logic, so it is not right to write tests on service method but store it in cotroller -- Hence, you did it right, and tests you have added need to be addresed to service (You can simply move them with small refactoring )

File: src/main/java/com/gachigang/ontherun/controller/DepartmentController.java (hunk 4 above)
Comment (harpestes) [reply src/main/java/com/gachigang/ontherun/controller/DepartmentController.java]: Done

File: src/test/java/com/gachigang/ontherun/controller/DepartmentControllerTest.java (hunk 6 above)
Comment (harpestes) [reply src/test/java/com/gachigang/ontherun/controller/DepartmentControllerTest.java]: I've rework it

Comment (harpestes) [reply src/test/java/com/gachigang/ontherun/controller/DepartmentControllerTest.java]: I've remove it

File: src/main/java/com/gachigang/ontherun/service/DepartmentService.java [hunk 7]
```diff
@@ -1,25 +1,43 @@
...
 
     @Transactional(readOnly = true)
     public Department getDepartmentByUserId(@NonNull final Long id) {
         return departmentRepository.findDepartmentByUsersContains(userRepository.findById(id)
                 .orElseThrow(NotFoundException::new));
     }
+
+    @Transactional(readOnly = true)
+    public Set<DepartmentDto> getDepartmentsByBusinessId(@NonNull Long businessId) {
+        Set<Department> departments = departmentRepository.findDepartmentsByBusiness(businessRepository.findById(businessId)
```
Comment (AlexandrVikhristyk) [inline src/main/java/com/gachigang/ontherun/service/DepartmentService.java:34]: Hmmmmm as far as I remember, all endoints witch return Collection of objects shoud return an empty collection instead of throwing error. It is way more efficient and handy to use from angular perspective. Even, repositpry layer should return emtry set if there is no departments

File: src/main/java/com/gachigang/ontherun/service/DepartmentService.java [hunk 8]
```diff
@@ -1,25 +1,43 @@
...
+                .orElseThrow(NotFoundException::new));
+        Set<DepartmentDto> resultDepartments = new HashSet<>();
+        DepartmentMapperImpl departmentMapper = new DepartmentMapperImpl();
```
Comment (AlexandrVikhristyk) [inline src/main/java/com/gachigang/ontherun/service/DepartmentService.java:37]: IoC?

File: src/main/java/com/gachigang/ontherun/service/DepartmentService.java [hunk 9]
```diff
@@ -1,25 +1,43 @@
...
+        for (Department d: departments) {
+            resultDepartments.add(departmentMapper.toDto(d));
```
Comment (AlexandrVikhristyk) [inline src/main/java/com/gachigang/ontherun/service/DepartmentService.java:39]: We should have toDtos method in mapper istead of running loop here

File: src/test/java/com/gachigang/ontherun/service/DepartmentServiceTest.java [hunk 10]
```diff
@@ -58,5 +68,46 @@ public void testGetDepartmentByUserIdNonExistingUser() {
...
+    public void testGetDepartmentsByBusinessId_Success() {
+        Business business = Business.builder()
+                .id(1L)
+                .build();
+        Department department1 = Department.builder()
+                .id(1L)
+                .users(Collections.emptySet())
+                .business(business)
+                .build();
+        Department department2= Department.builder()
```
Comment (AlexandrVikhristyk) [inline src/test/java/com/gachigang/ontherun/service/DepartmentServiceTest.java:82]: Add space after name

File: src/test/java/com/gachigang/ontherun/service/DepartmentServiceTest.java [hunk 11]
```diff
@@ -58,5 +68,46 @@ public void testGetDepartmentByUserIdNonExistingUser() {
...
+    public void testGetDepartmentsByBusinessId_InvalidBusinessId() {
+        Long invalidBusinessId = 2L;
+        when(businessRepository.findById(invalidBusinessId)).thenReturn(Optional.empty());
+
+        assertThrows(NotFoundException.class, () -> departmentService.getDepartmentsByBusinessId(invalidBusinessId));
+    }
+
+    @Test
+    public void testGetDepartmentsByBusinessId_NoDepartments() {
+        assertThrows(RuntimeException.class, () -> departmentService.getDepartmentsByBusinessId(1L));
```
Comment (AlexandrVikhristyk) [inline src/test/java/com/gachigang/ontherun/service/DepartmentServiceTest.java:110]: You throw NotFoundException, we should assert on the same condition

**Review by AlexandrVikhristyk — COMMENTED** [review]
Couple of comments

File: src/main/java/com/gachigang/ontherun/model/dto/DepartmentDto.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+import java.util.Set;
+
+/**
+ *  DTO class to represent {@link Department}.
+ */
+@Data
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class DepartmentDto {
```
Comment (AlexandrVikhristyk) [inline src/main/java/com/gachigang/ontherun/model/dto/DepartmentDto.java:20]: We have to use DTOs insted of entities (Only in dtos), because of recurcive fetching during whole transaction in Spring.
